src/OauthTokenStorage.php

```diff
@@ -0,0 +1,169 @@
+use Drupal\Core\Entity\EntityStorageException;
+use Drupal\key\KeyInterface;
+
+/**
+ * Storing and returning OAuth access token data.
+ */
+class OauthTokenStorage implements OauthTokenStorageInterface {
+
+  /**
+   * The timestamp when the token expires.
+   *
+   * This value is calculated from "expires_in" value which is the lifetime of the access token in seconds.
+   *
+   * @var int
+   */
+  protected $expires = 0;
+
+  /**
+   * Number of seconds extracted from token's expiration date when hasExpired() calculates.
+   *
+   * This ensures that token gets refreshed earlier than it expires.
+   *
+   * @var int
+   */
+  protected $leeway = 30;
+
+  /**
+   * The OAuth token key entity.
+   *
+   * @var \Drupal\key\KeyInterface
+   */
+  protected $key;
+
+  /**
+   * The key type of the OAuth token key entity.
+   *
+   * @var \Drupal\apigee_edge\Plugin\EdgeOauthTokenKeyTypeInterface
+   */
+  protected $keyType;
+
+  /**
+   * OauthTokenStorage constructor.
+   *
+   * @param \Drupal\key\KeyInterface $key
+   *   The OAuth token key entity.
+   */
+  public function __construct(KeyInterface $key) {
+    if (!($key->getKeyType() instanceof EdgeOauthTokenKeyTypeInterface)) {
+      throw new \InvalidArgumentException("Type of {$key->id()} key does not implement EdgeOauthTokenKeyTypeInterface.");
+    }
+
+    $this->key = $key;
+    $this->keyType = $key->getKeyType();
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getAccessToken(): ?string {
+    return $this->keyType->getAccessToken($this->key);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getTokenType(): ?string {
+    return $this->keyType->getTokenType($this->key);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getExpires(): int {
+    return $this->expires;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function hasExpired(): bool {
+    if ($this->getExpires() !== 0 && ($this->getExpires() - $this->leeway) > time()) {
+      return FALSE;
+    }
+
+    return TRUE;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function markExpired(): void {
+    $this->expires = 0;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getRefreshToken(): ?string {
+    return $this->keyType->getRefreshToken($this->key);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getScope(): string {
+    return $this->keyType->getScope($this->key);
+  }
+
+  /**
+   * {@inheritdoc}
+   *
+   * @throws \Apigee\Edge\Exception\OauthAuthenticationException
+   *   Thrown when the OAuth data cannot be saved.
+   */
+  public function saveToken(array $data): void {
+    try {
+      $this->key->setKeyValue($this->keyType->serialize($data));
+      $this->key->save();
+    }
+    catch (EntityStorageException $exception) {
+      throw new OauthAuthenticationException('Could not save the OAuth token data.');
+    }
+
+    if ($this->keyType->getExpiresIn($this->key) !== NULL && $this->keyType->getExpiresIn($this->key) !== 0) {
```

Comment:
Are you saving the calculated expiration date (unix timestamp) to the token storage this way? Because it should be saved, you can not use `expires_in` (number of seconds while the access token is valid from the time when it has been generated) after this object got destructed because the current page request has ended.